### DISCUSSION TITLE
optimization vs. linerarization

### DISCUSSION THREAD

**Discussion by andrisi** [pdfcpu/pdfcpu#578] [Ideas]
I wonder if optimization (as done by pdfcpu or [pdfsizeopt](https://github.com/pts/pdfsizeopt)) is in conflict with linerarization done by qpdf. If I optimize a file, and then try to linearize it, it gets big again (with pdfsizeopt). Could someone more knowledgable enlighten me? Or it is in conflict, so I need to chose which is better for a given file, linearization, so it can be range fetched from a webserver, or optimization, so its small, and there is really no need to linearize and depend on range fetching to speed things up. Thanks!

**Comment by petervwyatt** [comment]
Size optimization is a very different concept to linearization:

* Linearization (a.k.a. "fast web view") is about adding additional optional data structures to the start of a PDF file to enable the file to be processed without having to download the entire PDF (since standard PDF processing always starts at EOF). Think of it as displaying page 1 before pages 2-99 have even been downloaded. But adding such additional data will **always** make a file slightly bigger. With today's internet speeds, linearization of small PDFs is no longer worth it - and a significant number of PDF processors also do not support linearization.

* Optimization is about changing the internal PDF data structures to make the overall file size smaller. This can include using more or different compression algorithms, identifying reuse and removing duplicate objects, using more compact syntax, using compressed object streams and cross-reference streams from later versions of PDF, consilidating incremental updates, etc. Optimization may also remove dead (unreferenced) objects or other data it determines is not really needed - which may include the optional linearization data.

**Comment by andrisi** [comment]
Thanks for the exlanation @petervwyatt I understand the basics, the question is how they can be combined. I have lots of files to test on. So there is a 10mb file, which is reduced the 600K by optimization, but becomes 10mb again with linearization. I'd imagine that some optimizations remain after linearizing, but maybe not, and maybe qpdf just decompresses some things again.

**Comment by jberkenbilt** [comment]
qpdf author here...I have made some comments in qpdf/qpdf#919 but my position is consistent with the one presented here. I have described some pathological cases where qpdf could make a file very large by linearizing, but I'd have to see an example. I agree that linearizing small files is not worth it and, in general, there are fewer and fewer use cases for linearization these days.

**Comment by andrisi** [comment]
Will collect a batch of test files, and get back here. I'm working with lots of them, and lots of large ones (100-300mb+) where both optimization and linearization would come handy.

**Comment by hhrutter** [comment]
Thanks for the discussion.

I am closing this as answered/stale. Optimization and linearization solve different problems: optimization tries to reduce or normalize the internal PDF representation, while linearization adds optional structures for byte-range loading / fast web view and may therefore increase file size.

For pdfcpu, this remains a usage tradeoff rather than an actionable bug report. If somebody still sees a problematic size regression with current tooling, please open a new issue with:

- the input PDF or a reduced reproducible sample
- the exact pdfcpu/qpdf commands used
- file sizes before and after each step
- the expected result

Without concrete sample files there is nothing actionable here.
